rse-py-package.Rmd

```diff
@@ -0,0 +1,1460 @@
+quickly becomes essential.
+
+
+### Exercises: Examining the virtual environment 
+
+1. If you haven't already, create a `test` virtual environment and activate it.  Your prompt should now start with `(test)`.
+2. Run `pip list` -- what packages are installed?
+3. Run the `showpath.py` script from the earlier section, and note that the paths are different from your 
+base environment where you looked for packages.  They probably look like `/Users/pterry/envs/test/lib/python-3.7/site-packages`.
+4. Pick a package you know you have installed in your base environment -- pick `pytest` if you can't think of one -- and
+try to import it in a python interpreter.  Can it be found?
+5. Deactive the environment and try again.
+
+### Exercises: Installing a package into the virtual environment
+
+1. Re-activate the `test` virtual environment from above, and if necessary,
+re-run the `showpath.py` script.
+2. Let's install a simple package, `click`; it's a commonly used package for
+creating simple command-line tools, but for our purposes it's just a small
+self-contained package.  Run `pip install click`.
+3. Search in the directories output from `showpath.py` for a package named `click`;
+it will be a directory containing an `__init__.py`.
+4. Start a python interpreter, and run `import click`; it is installed in the environment.
+5. From that same interpretter, run `print(click.path)`; that is the path to the click package.
+It should report the same directory that you found it in.
+6. Exit the python interpretter
+
+Now let's just clarify that the _environment_ is different than the _directory_ you
+installed the virtual environment contents in:
+
+1. Start another terminal, and go to the directory containing the `test` virtual environment contents.
+2. Run the `showpath.py` script - does it show the directories in the base environment or the virtual environment?
+3. Start a python interpretter, and try `import click`.
+4.
+    a. If you didn't have `click` installed in the base environment, this will fail.
+    b. If you did have `click` installed in the base environment, this will be successful, but the click path will be different: `print(click.__path__)` to verify.
+5. Exit the python interpreter, and deactivate the `test` environment in the other terminal.
+
+## How can I create an installable Python package? {#rse-py-package-package}
+
+People can always get your package by cloning your repository and copying files from that
+(assuming your repository is accessible,
+which is should be for published research),
+but it's much friendlier to create something they can install
+(and it will be easier for us when it comes to running test suites
+and distributing, too!)
+
+For historical reasons,
+Python has several ways to build an installable package.
+We will show how to use [setuptools][setuptools],
+which is the lowest common denominator;
+[conda][conda] is a modern does-everything solution,
+but has larger startup overhead.
```

Comment:
"learning" rather than "startup" to emphasize it isn't a matter of runtime speed? (I'm being very pedantic...)

Reply:
Clarified the least common denominator part instead

Reply:
Rewrote to focus on lowest-common-denominator part